sys/linux/socket_nvme_of_tcp.txt

```diff
@@ -0,0 +1,623 @@
+
+#########################################
+### nvme_tcp_data_pdu ###
+
+nvme_tcp_data_pdu {
+	hdr		nvme_tcp_hdr_data_pdu_union
+	command_id	int16
+	ttag		int16
+	data_offset	int32
+	data_length	int32
+	rsvd		array[int8, 4]
+} [size[NVME_TCP_DATA_PDU_SIZE]]
+
+nvme_tcp_hdr_data_pdu_union [
+	nvme_tcp_hdr_data_pdu_h2c		nvme_tcp_hdr_t_const[nvme_tcp_h2c_data, const[NVME_TCP_DATA_PDU_SIZE, int8]]
+	nvme_tcp_hdr_data_pdu_h2c_hdigest	nvme_tcp_hdr_t_const_extra[nvme_tcp_h2c_data, const[NVME_TCP_DATA_PDU_SIZE, int8], const[NVME_TCP_HDR_H2C_DIGEST_PDO_HDIGEST, int8]]
+	nvme_tcp_hdr_data_pdu_h2c_no_hdigest	nvme_tcp_hdr_t_const_extra[nvme_tcp_h2c_data, const[NVME_TCP_DATA_PDU_SIZE, int8], const[NVME_TCP_HDR_H2C_DIGEST_PDO_NO_HDIGEST, int8]]
+	nvme_tcp_hdr_data_pdu_c2h		nvme_tcp_hdr_t_const[nvme_tcp_c2h_data, const[0, int8]]
+]
+
+define NVME_TCP_HDR_H2C_DIGEST_PDO_HDIGEST	NVME_TCP_DATA_PDU_SIZE + NVME_TCP_DIGEST_LENGTH
+define NVME_TCP_HDR_H2C_DIGEST_PDO_NO_HDIGEST	NVME_TCP_DATA_PDU_SIZE
+define NVME_TCP_HDR_H2C_DIGEST_PLEN_HDIGEST_DDIGEST	NVME_TCP_HDR_H2C_DIGEST_PDO_HDIGEST + NVME_TCP_DIGEST_LENGTH
+
+### End of nvme_tcp_r2t_pdu ###
+#########################################
+
+#########################################
+### nvme_tcp_r2t_pdu ###
+
+nvme_tcp_r2t_pdu {
+	hdr		nvme_tcp_hdr_r2t_pdu
+	command_id	int16
+	ttag		int16
+	r2t_offset	int32
+	r2t_length	int32
+	rsvd		array[int8, 4]
+} [packed]
+
+type nvme_tcp_hdr_r2t_pdu nvme_tcp_hdr_t_const[nvme_tcp_r2t, const[0, int8]]
+
+### End of nvme_tcp_r2t_pdu ###
+#########################################
+
+#########################################
+### nvme_tcp_rsp_pdu ###
+
+nvme_tcp_rsp_pdu {
+	hdr	nvme_tcp_hdr_rsp_pdu
+	cqe	nvme_completion
+} [packed]
+
+nvme_completion {
+	result		nvme_result
+	sq_head		int16
+	sq_id		int16
+	command_id	int16
+	status		flags[nvme_tcp_fatal_error_status, int16]
+} [packed]
+
+nvme_result [
+	u16	int16
+	u32	int32
+	u64	int64
+]
+
+type nvme_tcp_hdr_rsp_pdu nvme_tcp_hdr_t_const[nvme_tcp_rsp, const[0, int8]]
+
+### End of nvme_tcp_rsp_pdu ###
+#########################################
+
+#########################################
+### nvme_tcp_icresp_pdu ###
+
+nvme_tcp_icresp_pdu {
+	hdr	nvme_tcp_hdr_icresp_pdu
+	pfv	flags[nvme_tcp_pfv, int16]
+# Linux NVMe driver only support cpda = 0
+	cpda	const[0, int8]
+	digest	flags[nvme_tcp_digest_option, int8]
+	maxdata	int32
+	rsvd	array[int8, 112]
+} [packed]
+
+type nvme_tcp_hdr_icresp_pdu nvme_tcp_hdr_t_const[nvme_tcp_icresp, const[0, int8]]
+
+### End of nvme_tcp_icresp_pdu ###
+#########################################
+
+#########################################
+### nvme_tcp_icreq_pdu ###
+
+nvme_tcp_icreq_pdu {
+	hdr	nvme_tcp_hdr_icreq_pdu_union
+	pfv	flags[nvme_tcp_pfv, int16]
+# Linux NVMe driver only support hcpda = 0
+	hpda	const[0, int8]
+	digest	flags[nvme_tcp_digest_option, int8]
+	maxr2t	int32
+	rsvd	array[int8, 112]
+} [size[NVME_TCP_ICREQ_PDU_SIZE]]
+
+nvme_tcp_hdr_icreq_pdu_union [
```

Comment:
> but in case sometime in the future there will be another layout of the header

You might as well be able to incorporate those future changes (if they ever happpen) directly into `nvme_tcp_hdr_t_const_plen`. Or, in the worst case, we can re-introduce that union at any moment. For now it just complicates the descriptions for no benefit.